File: claude-plugins/web-mcp-server/skills/webmcp-extension-websocket-protocol/scripts/webmcp_harness.py
#!/usr/bin/env python3
import asyncio
import json
import sys
import uuid
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

try:
    import websockets
except ImportError as exc:
    raise SystemExit(
        "Missing dependency: websockets. Install with: pip install -r requirements.txt"
    ) from exc

Command = str
JsonObject = Dict[str, Any]


@dataclass
class GlobalOptions:
    port: int = 8765
    timeout_ms: int = 70000
    session_id: Optional[str] = None
    pretty: bool = False


@dataclass
class ParsedArgs:
    command: Command
    options: GlobalOptions
    url: Optional[str]
    focus: bool
    tab_id: Optional[int]
    tool_name: Optional[str]
    tool_args: JsonObject


def parse_bool(value: str, flag: str) -> bool:
    if value == "true":
        return True
    if value == "false":
        return False
    raise ValueError(f"{flag} must be true or false")


def parse_int(value: str, flag: str) -> int:
    try:
        return int(value)
    except ValueError as exc:
        raise ValueError(f"{flag} must be an integer") from exc


def require_value(argv: List[str], index: int, flag: str) -> str:
    if index + 1 >= len(argv):
        raise ValueError(f"{flag} requires a value")
    return argv[index + 1]


def parse_args(argv: List[str]) -> ParsedArgs:
    options = GlobalOptions()
    command: Optional[Command] = None
    url: Optional[str] = None
    focus = True
    tab_id: Optional[int] = None
    tool_name: Optional[str] = None
    tool_args: JsonObject = {}

    i = 0
    while i < len(argv):
        token = argv[i]

        if token.startswith("--"):
            if token == "--pretty":
                options.pretty = True
                i += 1
                continue

            value = require_value(argv, i, token)
            if token == "--port":
                options.port = parse_int(value, token)
            elif token == "--timeout-ms":
                options.timeout_ms = parse_int(value, token)
            elif token == "--session-id":
                options.session_id = value
            elif token == "--url":
                url = value
            elif token == "--focus":
                focus = parse_bool(value, token)
            elif token == "--tab-id":
                tab_id = parse_int(value, token)
            elif token == "--tool-name":
                tool_name = value
            elif token == "--args-json":
                try:
                    parsed = json.loads(value)
                except json.JSONDecodeError as exc:
                    raise ValueError("--args-json must be valid JSON") from exc
                if not isinstance(parsed, dict):
                    raise ValueError("--args-json must be a JSON object")
                tool_args = parsed
            else:
                raise ValueError(f"Unknown flag: {token}")

            i += 2
            continue

        if command is None:
            if token in {"list-tabs", "open-tab", "discover-tools", "call-tool"}:
                command = token
            else:
                raise ValueError(f"Unknown command: {token}")
            i += 1
            continue

        raise ValueError(f"Unexpected positional argument: {token}")

    if command is None:
        raise ValueError(
            "Missing command. Use one of: list-tabs, open-tab, discover-tools, call-tool"
        )

    if options.port < 8765 or options.port > 8785:
        raise ValueError("--port must be in range 8765-8785")

    if options.timeout_ms <= 0:
        raise ValueError("--timeout-ms must be > 0")

    if command == "open-tab" and not url:
        raise ValueError("open-tab requires --url <url>")

    if command in {"discover-tools", "call-tool"} and tab_id is None:
        raise ValueError(f"{command} requires --tab-id <id>")

    if command == "call-tool" and not tool_name:
        raise ValueError("call-tool requires --tool-name <name>")

    return ParsedArgs(
        command=command,
        options=options,
        url=url,
        focus=focus,
        tab_id=tab_id,
        tool_name=tool_name,
        tool_args=tool_args,
    )


def merge_session_id(payload: JsonObject, session_id: Optional[str]) -> JsonObject:
    if session_id is None:
        return payload
    merged = dict(payload)
    merged["sessionId"] = session_id
    return merged


class Harness:
    def __init__(self, options: GlobalOptions):
        self.options = options
        self.server = None
        self.socket = None
        self.connection_future: Optional[asyncio.Future[None]] = None
        self.connect_future: Optional[asyncio.Future[JsonObject]] = None
        self.pending_calls: Dict[str, asyncio.Future[Any]] = {}
        self.pending_requests: Dict[str, asyncio.Future[JsonObject]] = {}
        self.tabs: Dict[int, JsonObject] = {}
        self.browser: Optional[JsonObject] = None

    @property
    def timeout_seconds(self) -> float:
        return self.options.timeout_ms / 1000

    async def start(self) -> None:
        self.connection_future = asyncio.get_running_loop().create_future()
        try:
            self.server = await websockets.serve(
                self._handle_connection, "localhost", self.options.port
            )
        except OSError as exc:
            if exc.errno in {48, 98, 10048}:
                raise RuntimeError(
                    f"Port {self.options.port} is already in use. Choose another port in the allowed range 8765-8785."
                ) from exc
            raise
        await asyncio.wait_for(self.connection_future, timeout=self.timeout_seconds)

    async def _handle_connection(self, websocket) -> None:
        if self.socket is not None:
            await websocket.close(code=1000, reason="Already connected")
            return

        self.socket = websocket
        if self.connection_future and not self.connection_future.done():
            self.connection_future.set_result(None)

        try:
            async for raw in websocket:
                if isinstance(raw, bytes):
                    text = raw.decode("utf-8", errors="ignore")
                else:
                    text = raw
                await self._handle_message(text)
        except Exception:
            pass
        finally:
            if self.socket is websocket:
                self.socket = None
                self._reject_all(RuntimeError("Extension disconnected"))

    async def _handle_message(self, raw: str) -> None:
        try:
            message = json.loads(raw)
        except json.JSONDecodeError:
            return

        if not isinstance(message, dict):
            return

        msg_type = message.get("type")
        if not isinstance(msg_type, str):
            return

        if msg_type == "ping":
            await self._safe_send({"type": "pong"})
            return

        if msg_type == "connected":
            browser = message.get("browser")
            tabs = message.get("tabs")
            if isinstance(browser, dict):
                self.browser = browser
            self.tabs.clear()
            if isinstance(tabs, list):
                for tab in tabs:
                    if isinstance(tab, dict) and isinstance(tab.get("id"), int):
                        self.tabs[tab["id"]] = tab

            if self.connect_future and not self.connect_future.done():
                self.connect_future.set_result(
                    {
                        "browser": self.browser or {},
                        "tabs": self.list_tabs(),
                    }
                )
            return

        if msg_type == "tabCreated":
            tab = message.get("tab")
            if isinstance(tab, dict) and isinstance(tab.get("id"), int):
                self.tabs[tab["id"]] = tab
            request_id = message.get("requestId")
            if isinstance(request_id, str):
                future = self.pending_requests.get(request_id)
                if future and not future.done():
                    future.set_result(tab if isinstance(tab, dict) else {})
            return

        if msg_type == "tabFocused":
            tab_id = message.get("tabId")
            if isinstance(tab_id, int):
                existing = self.tabs.get(tab_id, {"id": tab_id})
                tools = message.get("tools")
                merged = dict(existing)
                if isinstance(tools, list):
                    merged["tools"] = tools
                self.tabs[tab_id] = merged

            request_id = message.get("requestId")
            if isinstance(request_id, str):
                future = self.pending_requests.get(request_id)
                if future and not future.done():
                    resolved_tab = self.tabs.get(tab_id, {"id": tab_id}) if isinstance(tab_id, int) else {}
                    future.set_result(resolved_tab)
            return

        if msg_type == "tabUpdated":
            tab = message.get("tab")
            if isinstance(tab, dict) and isinstance(tab.get("id"), int):
                self.tabs[tab["id"]] = tab
            return

        if msg_type == "tabClosed":
            tab_id = message.get("tabId")
            if isinstance(tab_id, int):
                self.tabs.pop(tab_id, None)
            return

        if msg_type == "toolsChanged":
            tab_id = message.get("tabId")
            tools = message.get("tools")
            if isinstance(tab_id, int):
                existing = self.tabs.get(tab_id, {"id": tab_id})
                merged = dict(existing)
                merged["tools"] = tools if isinstance(tools, list) else []
                self.tabs[tab_id] = merged
            return

        if msg_type == "toolsDiscovered":
            call_id = message.get("callId")
            if isinstance(call_id, str):
                future = self.pending_calls.get(call_id)
                if future and not future.done():
                    future.set_result(
                        {
                            "tabId": message.get("tabId"),
                            "tools": message.get("tools") if isinstance(message.get("tools"), list) else [],
                        }
                    )
            return

        if msg_type == "toolResult":
            call_id = message.get("callId")
            if isinstance(call_id, str):
                future = self.pending_calls.get(call_id)
                if future and not future.done():
                    error = message.get("error")
                    if isinstance(error, str):
                        future.set_exception(RuntimeError(error))
                    else:
                        future.set_result(message.get("result"))
            return

        if msg_type == "disconnected":
            self._reject_all(RuntimeError("Extension sent disconnected event"))

    async def ensure_connected_snapshot(self) -> JsonObject:
        if self.browser is not None:
            return {"browser": self.browser, "tabs": self.list_tabs()}

        if self.socket is None:
            raise RuntimeError("Extension socket is not connected")

        if self.connect_future is None:
            self.connect_future = asyncio.get_running_loop().create_future()
            await self._send(merge_session_id({"type": "connect"}, self.options.session_id))

        try:
            return await asyncio.wait_for(self.connect_future, timeout=self.timeout_seconds)
        except asyncio.TimeoutError as exc:
            self.connect_future = None
            raise RuntimeError(f"connect timed out after {self.options.timeout_ms}ms") from exc
        finally:
            if self.connect_future and self.connect_future.done():
                self.connect_future = None

    def list_tabs(self) -> List[JsonObject]:
        return list(self.tabs.values())

    async def open_tab(self, url: str, focus: bool) -> JsonObject:
        await self.ensure_connected_snapshot()

        request_id = f"request_{uuid.uuid4()}"
        future = asyncio.get_running_loop().create_future()
        self.pending_requests[request_id] = future

        await self._send(
            merge_session_id(
                {
                    "type": "openTab",
                    "url": url,
                    "focus": focus,
                    "requestId": request_id,
                },
                self.options.session_id,
            )
        )

        try:
            return await asyncio.wait_for(future, timeout=self.timeout_seconds)
        except asyncio.TimeoutError as exc:
            raise RuntimeError(f"openTab timed out after {self.options.timeout_ms}ms") from exc
        finally:
            self.pending_requests.pop(request_id, None)

    async def discover_tools(self, tab_id: int) -> JsonObject:
        await self.ensure_connected_snapshot()

        call_id = f"call_{uuid.uuid4()}"
        future = asyncio.get_running_loop().create_future()
        self.pending_calls[call_id] = future

        await self._send(
            merge_session_id(
                {
                    "type": "discoverTools",
                    "callId": call_id,
                    "tabId": tab_id,
                },
                self.options.session_id,
            )
        )

        try:
            result = await asyncio.wait_for(future, timeout=self.timeout_seconds)
            return result if isinstance(result, dict) else {"tabId": tab_id, "tools": []}
        except asyncio.TimeoutError as exc:
            raise RuntimeError(f"discoverTools timed out after {self.options.timeout_ms}ms") from exc
        finally:
            self.pending_calls.pop(call_id, None)

    async def call_tool(self, tab_id: int, tool_name: str, args: JsonObject) -> Any:
        await self.ensure_connected_snapshot()

        call_id = f"call_{uuid.uuid4()}"
        future = asyncio.get_running_loop().create_future()
        self.pending_calls[call_id] = future

        await self._send(
            merge_session_id(
                {
                    "type": "callTool",
                    "callId": call_id,
                    "tabId": tab_id,
                    "toolName": tool_name,
                    "args": args,
                },
                self.options.session_id,
            )
        )

        try:
            return await asyncio.wait_for(future, timeout=self.timeout_seconds)
        except asyncio.TimeoutError as exc:
            raise RuntimeError(f"callTool timed out after {self.options.timeout_ms}ms") from exc
        finally:
            self.pending_calls.pop(call_id, None)

    async def _send(self, payload: JsonObject) -> None:
        if self.socket is None:
            raise RuntimeError("Extension socket is not connected")
        await self.socket.send(json.dumps(payload))

    async def _safe_send(self, payload: JsonObject) -> None:
        try:
            if self.socket is not None:
                await self.socket.send(json.dumps(payload))
        except Exception:
            pass

    def _reject_all(self, error: Exception) -> None:
        if self.connect_future and not self.connect_future.done():
            self.connect_future.set_exception(error)

        for key, future in list(self.pending_calls.items()):
            if not future.done():
                future.set_exception(error)
            self.pending_calls.pop(key, None)

        for key, future in list(self.pending_requests.items()):
            if not future.done():
                future.set_exception(error)
            self.pending_requests.pop(key, None)

        if self.connection_future and not self.connection_future.done():
            self.connection_future.set_exception(error)

    async def close(self) -> None:
        self._reject_all(RuntimeError("Harness shutting down"))

        if self.socket is not None:
            try:
                await self.socket.close()
            except Exception:
                pass
            self.socket = None

        if self.server is not None:
            self.server.close()
            await self.server.wait_closed()
            self.server = None


def write_json(payload: JsonObject, pretty: bool) -> None:
    if pretty:
        print(json.dumps(payload, indent=2))
    else:
        print(json.dumps(payload, separators=(",", ":")))


async def run(argv: List[str]) -> int:
    try:
        parsed = parse_args(argv)
    except Exception as error:
        write_json({"ok": False, "error": str(error)}, True)
        return 1

    harness = Harness(parsed.options)

    try:
        await harness.start()
        snapshot = await harness.ensure_connected_snapshot()

        if parsed.command == "list-tabs":
            write_json(
                {
                    "ok": True,
                    "port": parsed.options.port,
                    "browser": snapshot.get("browser"),
                    "tabs": harness.list_tabs(),
                },
                parsed.options.pretty,
            )
            return 0

        if parsed.command == "open-tab":
            tab = await harness.open_tab(parsed.url or "", parsed.focus)
            write_json({"ok": True, "tab": tab}, parsed.options.pretty)
            return 0

        if parsed.command == "discover-tools":
            discovered = await harness.discover_tools(parsed.tab_id or 0)
            write_json(
                {
                    "ok": True,
                    "tabId": parsed.tab_id,
                    "tools": discovered.get("tools", []),
                },
                parsed.options.pretty,
            )
            return 0

        if parsed.command == "call-tool":
            result = await harness.call_tool(
                parsed.tab_id or 0,
                parsed.tool_name or "",
                parsed.tool_args,
            )
            write_json(
                {
                    "ok": True,
                    "tabId": parsed.tab_id,
                    "toolName": parsed.tool_name,
                    "result": result,
                },
                parsed.options.pretty,
            )
            return 0

        write_json({"ok": False, "error": f"Unknown command: {parsed.command}"}, parsed.options.pretty)
        return 1
    except Exception as error:
        write_json({"ok": False, "error": str(error)}, parsed.options.pretty)
        return 1
    finally:
        await harness.close()


def main() -> None:
    exit_code = asyncio.run(run(sys.argv[1:]))
    raise SystemExit(exit_code)


if __name__ == "__main__":
    main()

File: claude-plugins/web-mcp-server/skills/webmcp-extension-websocket-protocol/scripts/webmcp_harness.ts
import {randomUUID} from 'node:crypto';
import {WebSocket, WebSocketServer} from 'ws';

type Command = 'list-tabs' | 'open-tab' | 'discover-tools' | 'call-tool';

type JsonObject = Record<string, unknown>;

interface GlobalOptions {
  port: number;
  timeoutMs: number;
  sessionId?: string;
  pretty: boolean;
}

interface ParsedArgs {
  command: Command;
  options: GlobalOptions;
  url?: string;
  focus: boolean;
  tabId?: number;
  toolName?: string;
  toolArgs: Record<string, unknown>;
}

interface PendingState<T> {
  resolve: (value: T) => void;
  reject: (error: Error) => void;
  timeout: ReturnType<typeof setTimeout>;
}

interface ConnectedSnapshot {
  browser: JsonObject;
  tabs: JsonObject[];
}

function parseBoolean(value: string, flag: string): boolean {
  if (value === 'true') {
    return true;
  }
  if (value === 'false') {
    return false;
  }

  throw new Error(`${flag} must be true or false`);
}

function parseInteger(value: string, flag: string): number {
  const parsed = Number.parseInt(value, 10);
  if (Number.isNaN(parsed)) {
    throw new Error(`${flag} must be an integer`);
  }
  return parsed;
}

function requireValue(argv: string[], index: number, flag: string): string {
  const value = argv[index + 1];
  if (value === undefined) {
    throw new Error(`${flag} requires a value`);
  }
  return value;
}

function parseArgs(argv: string[]): ParsedArgs {
  const options: GlobalOptions = {
    port: 8765,
    timeoutMs: 70_000,
    pretty: false
  };

  let command: Command | null = null;
  let url: string | undefined;
  let focus = true;
  let tabId: number | undefined;
  let toolName: string | undefined;
  let toolArgs: Record<string, unknown> = {};

  let i = 0;
  while (i < argv.length) {
    const token = argv[i];

    if (token.startsWith('--')) {
      if (token === '--pretty') {
        options.pretty = true;
        i += 1;
        continue;
      }

      const value = requireValue(argv, i, token);
      switch (token) {
        case '--port':
          options.port = parseInteger(value, token);
          break;
        case '--timeout-ms':
          options.timeoutMs = parseInteger(value, token);
          break;
        case '--session-id':
          options.sessionId = value;
          break;
        case '--url':
          url = value;
          break;
        case '--focus':
          focus = parseBoolean(value, token);
          break;
        case '--tab-id':
          tabId = parseInteger(value, token);
          break;
        case '--tool-name':
          toolName = value;
          break;
        case '--args-json': {
          let parsed: unknown;
          try {
            parsed = JSON.parse(value);
          } catch {
            throw new Error('--args-json must be valid JSON');
          }

          if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
            throw new Error('--args-json must be a JSON object');
          }

          toolArgs = parsed as Record<string, unknown>;
          break;
        }
        default:
          throw new Error(`Unknown flag: ${token}`);
      }

      i += 2;
      continue;
    }

    if (command === null) {
      if (
        token === 'list-tabs' ||
        token === 'open-tab' ||
        token === 'discover-tools' ||
        token === 'call-tool'
      ) {
        command = token;
      } else {
        throw new Error(`Unknown command: ${token}`);
      }
      i += 1;
      continue;
    }

    throw new Error(`Unexpected positional argument: ${token}`);
  }

  if (command === null) {
    throw new Error(
      'Missing command. Use one of: list-tabs, open-tab, discover-tools, call-tool'
    );
  }

  if (options.port < 8765 || options.port > 8785) {
    throw new Error('--port must be in range 8765-8785');
  }

  if (options.timeoutMs <= 0) {
    throw new Error('--timeout-ms must be > 0');
  }

  if (command === 'open-tab' && !url) {
    throw new Error('open-tab requires --url <url>');
  }

  if ((command === 'discover-tools' || command === 'call-tool') && tabId === undefined) {
    throw new Error(`${command} requires --tab-id <id>`);
  }

  if (command === 'call-tool' && !toolName) {
    throw new Error('call-tool requires --tool-name <name>');
  }

  return {
    command,
    options,
    url,
    focus,
    tabId,
    toolName,
    toolArgs
  };
}

function asObject(value: unknown): JsonObject | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null;
  }
  return value as JsonObject;
}

function asString(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

function asNumber(value: unknown): number | undefined {
  return typeof value === 'number' ? value : undefined;
}

function mergeSessionId(payload: JsonObject, sessionId?: string): JsonObject {
  if (!sessionId) {
    return payload;
  }

  return {
    ...payload,
    sessionId
  };
}

class Harness {
  private readonly options: GlobalOptions;
  private server: WebSocketServer | null = null;
  private socket: WebSocket | null = null;
  private connectionResolve: (() => void) | null = null;
  private connectionReject: ((error: Error) => void) | null = null;
  private readonly tabs = new Map<number, JsonObject>();
  private browser: JsonObject | null = null;

  private pendingConnect: {
    promise: Promise<ConnectedSnapshot>;
    state: PendingState<ConnectedSnapshot>;
  } | null = null;

  private readonly pendingCalls = new Map<string, PendingState<unknown>>();
  private readonly pendingRequests = new Map<string, PendingState<JsonObject>>();

  constructor(options: GlobalOptions) {
    this.options = options;
  }

  async start(): Promise<void> {
    await this.startServer();
    await this.waitForConnection();
  }

  private async startServer(): Promise<void> {
    if (this.server) {
      return;
    }

    const server = new WebSocketServer({port: this.options.port});
    this.server = server;

    await new Promise<void>((resolve, reject) => {
      const onError = (error: Error) => {
        server.off('listening', onListening);
        const code = (error as NodeJS.ErrnoException).code;
        if (code === 'EADDRINUSE') {
          reject(
            new Error(
              `Port ${this.options.port} is already in use. Choose another port in the allowed range 8765-8785.`
            )
          );
          return;
        }
        reject(error);
      };
      const onListening = () => {
        server.off('error', onError);
        resolve();
      };

      server.once('error', onError);
      server.once('listening', onListening);
    });

    server.on('connection', (socket) => {
      if (this.socket && this.socket.readyState === WebSocket.OPEN) {
        socket.close();
        return;
      }

      this.socket = socket;
      this.connectionResolve?.();
      this.connectionResolve = null;
      this.connectionReject = null;

      socket.on('message', (data) => {
        this.handleIncoming(data.toString());
      });

      socket.on('close', () => {
        if (this.socket === socket) {
          this.socket = null;
          this.rejectAllPending(new Error('Extension disconnected'));
        }
      });

      socket.on('error', () => {
        // "close" performs cleanup.
      });
    });
  }

  private async waitForConnection(): Promise<void> {
    if (this.socket && this.socket.readyState === WebSocket.OPEN) {
      return;
    }

    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.connectionResolve = null;
        this.connectionReject = null;
        reject(
          new Error(
            `Timed out waiting for extension connection on ws://localhost:${this.options.port}`
          )
        );
      }, this.options.timeoutMs);

      this.connectionResolve = () => {
        clearTimeout(timeout);
        resolve();
      };

      this.connectionReject = (error) => {
        clearTimeout(timeout);
        reject(error);
      };
    });
  }

  private handleIncoming(raw: string): void {
    let message: JsonObject;
    try {
      message = JSON.parse(raw) as JsonObject;
    } catch {
      return;
    }

    const type = asString(message.type);
    if (!type) {
      return;
    }

    if (type === 'ping') {
      this.safeSend({type: 'pong'});
      return;
    }

    switch (type) {
      case 'connected': {
        const browser = asObject(message.browser);
        const tabs = Array.isArray(message.tabs)
          ? (message.tabs.filter((tab) => asObject(tab) !== null) as JsonObject[])
          : [];

        if (browser) {
          this.browser = browser;
        }
        this.tabs.clear();
        for (const tab of tabs) {
          const tabId = asNumber(tab.id);
          if (tabId !== undefined) {
            this.tabs.set(tabId, tab);
          }
        }

        if (this.pendingConnect) {
          const {state} = this.pendingConnect;
          clearTimeout(state.timeout);
          this.pendingConnect = null;
          state.resolve({
            browser: this.browser ?? {},
            tabs: this.listTabs()
          });
        }
        break;
      }

      case 'tabCreated': {
        const tab = asObject(message.tab);
        const tabId = tab ? asNumber(tab.id) : undefined;
        if (tab && tabId !== undefined) {
          this.tabs.set(tabId, tab);
        }

        const requestId = asString(message.requestId);
        if (requestId) {
          const pending = this.pendingRequests.get(requestId);
          if (pending) {
            clearTimeout(pending.timeout);
            this.pendingRequests.delete(requestId);
            pending.resolve(tab ?? {});
          }
        }
        break;
      }

      case 'tabFocused': {
        const tabId = asNumber(message.tabId);
        if (tabId !== undefined) {
          const existing = this.tabs.get(tabId) ?? {id: tabId};
          const tools = message.tools;
          this.tabs.set(tabId, {
            ...existing,
            ...(Array.isArray(tools) ? {tools} : {})
          });
        }

        const requestId = asString(message.requestId);
        if (requestId) {
          const pending = this.pendingRequests.get(requestId);
          if (pending) {
            clearTimeout(pending.timeout);
            this.pendingRequests.delete(requestId);
            const tab = tabId !== undefined ? this.tabs.get(tabId) ?? {id: tabId} : {};
            pending.resolve(tab);
          }
        }
        break;
      }

      case 'tabUpdated': {
        const tab = asObject(message.tab);
        const tabId = tab ? asNumber(tab.id) : undefined;
        if (tab && tabId !== undefined) {
          this.tabs.set(tabId, tab);
        }
        break;
      }

      case 'tabClosed': {
        const tabId = asNumber(message.tabId);
        if (tabId !== undefined) {
          this.tabs.delete(tabId);
        }
        break;
      }

      case 'toolsChanged': {
        const tabId = asNumber(message.tabId);
        if (tabId !== undefined) {
          const existing = this.tabs.get(tabId) ?? {id: tabId};
          const tools = Array.isArray(message.tools) ? message.tools : [];
          this.tabs.set(tabId, {
            ...existing,
            tools
          });
        }
        break;
      }

      case 'toolsDiscovered': {
        const callId = asString(message.callId);
        if (!callId) {
          break;
        }

        const pending = this.pendingCalls.get(callId);
        if (!pending) {
          break;
        }

        clearTimeout(pending.timeout);
        this.pendingCalls.delete(callId);
        pending.resolve({
          tabId: message.tabId,
          tools: Array.isArray(message.tools) ? message.tools : []
        });
        break;
      }

      case 'toolResult': {
        const callId = asString(message.callId);
        if (!callId) {
          break;
        }

        const pending = this.pendingCalls.get(callId);
        if (!pending) {
          break;
        }

        clearTimeout(pending.timeout);
        this.pendingCalls.delete(callId);

        const error = asString(message.error);
        if (error) {
          pending.reject(new Error(error));
        } else {
          pending.resolve(message.result);
        }
        break;
      }

      case 'disconnected': {
        this.rejectAllPending(new Error('Extension sent disconnected event'));
        break;
      }
    }
  }

  async ensureConnectedSnapshot(): Promise<ConnectedSnapshot> {
    if (this.browser) {
      return {
        browser: this.browser,
        tabs: this.listTabs()
      };
    }

    const socket = this.socket;
    if (!socket || socket.readyState !== WebSocket.OPEN) {
      throw new Error('Extension socket is not connected');
    }

    if (this.pendingConnect) {
      return await this.pendingConnect.promise;
    }

    const pendingPromise = new Promise<ConnectedSnapshot>((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingConnect = null;
        reject(new Error(`connect timed out after ${this.options.timeoutMs}ms`));
      }, this.options.timeoutMs);

      this.pendingConnect = {
        promise: Promise.resolve({browser: {}, tabs: []}),
        state: {
          timeout,
          resolve,
          reject
        }
      };
    });

    if (this.pendingConnect) {
      this.pendingConnect.promise = pendingPromise;
    }

    this.send(mergeSessionId({type: 'connect'}, this.options.sessionId));

    try {
      return await pendingPromise;
    } catch (error) {
      if (this.pendingConnect) {
        clearTimeout(this.pendingConnect.state.timeout);
        this.pendingConnect = null;
      }
      throw error;
    }
  }

  async openTab(url: string, focus: boolean): Promise<JsonObject> {
    await this.ensureConnectedSnapshot();

    const requestId = `request_${randomUUID()}`;
    const response = this.createPendingInMap<JsonObject>(
      this.pendingRequests,
      requestId,
      'openTab'
    );

    this.send(
      mergeSessionId(
        {
          type: 'openTab',
          url,
          focus,
          requestId
        },
        this.options.sessionId
      )
    );

    return await response;
  }

  async discoverTools(tabId: number): Promise<{tabId: number; tools: unknown[]}> {
    await this.ensureConnectedSnapshot();

    const callId = `call_${randomUUID()}`;
    const response = this.createPendingInMap<{tabId: number; tools: unknown[]}>(
      this.pendingCalls,
      callId,
      'discoverTools'
    );

    this.send(
      mergeSessionId(
        {
          type: 'discoverTools',
          callId,
          tabId
        },
        this.options.sessionId
      )
    );

    return await response;
  }

  async callTool(
    tabId: number,
    toolName: string,
    args: Record<string, unknown>
  ): Promise<unknown> {
    await this.ensureConnectedSnapshot();

    const callId = `call_${randomUUID()}`;
    const response = this.createPendingInMap<unknown>(
      this.pendingCalls,
      callId,
      'callTool'
    );

    this.send(
      mergeSessionId(
        {
          type: 'callTool',
          callId,
          tabId,
          toolName,
          args
        },
        this.options.sessionId
      )
    );

    return await response;
  }

  listTabs(): JsonObject[] {
    return Array.from(this.tabs.values());
  }

  private createPendingInMap<T>(
    map: Map<string, PendingState<T>>,
    key: string,
    label: string
  ): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const timeout = setTimeout(() => {
        map.delete(key);
        reject(new Error(`${label} timed out after ${this.options.timeoutMs}ms`));
      }, this.options.timeoutMs);

      map.set(key, {
        timeout,
        resolve,
        reject
      });
    });
  }

  private send(message: JsonObject): void {
    if (!this.socket || this.socket.readyState !== WebSocket.OPEN) {
      throw new Error('Extension socket is not connected');
    }

    this.socket.send(JSON.stringify(message));
  }

  private safeSend(message: JsonObject): void {
    try {
      if (this.socket && this.socket.readyState === WebSocket.OPEN) {
        this.socket.send(JSON.stringify(message));
      }
    } catch {
      // ignore
    }
  }

  private rejectAllPending(error: Error): void {
    if (this.pendingConnect) {
      clearTimeout(this.pendingConnect.state.timeout);
      this.pendingConnect.state.reject(error);
      this.pendingConnect = null;
    }

    for (const [key, pending] of this.pendingCalls) {
      clearTimeout(pending.timeout);
      pending.reject(error);
      this.pendingCalls.delete(key);
    }

    for (const [key, pending] of this.pendingRequests) {
      clearTimeout(pending.timeout);
      pending.reject(error);
      this.pendingRequests.delete(key);
    }

    this.connectionReject?.(error);
    this.connectionReject = null;
    this.connectionResolve = null;
  }

  async close(): Promise<void> {
    this.rejectAllPending(new Error('Harness shutting down'));

    if (this.socket && this.socket.readyState === WebSocket.OPEN) {
      this.socket.close();
    }
    this.socket = null;

    if (!this.server) {
      return;
    }

    const server = this.server;
    this.server = null;

    await new Promise<void>((resolve) => {
      server.close(() => resolve());
    });
  }
}

function writeJson(payload: JsonObject, pretty: boolean): void {
  process.stdout.write(`${JSON.stringify(payload, null, pretty ? 2 : 0)}\n`);
}

async function run(): Promise<number> {
  let parsed: ParsedArgs;
  try {
    parsed = parseArgs(process.argv.slice(2));
  } catch (error) {
    writeJson(
      {
        ok: false,
        error: error instanceof Error ? error.message : String(error)
      },
      true
    );
    return 1;
  }

  const harness = new Harness(parsed.options);

  try {
    await harness.start();
    const snapshot = await harness.ensureConnectedSnapshot();

    switch (parsed.command) {
      case 'list-tabs': {
        writeJson(
          {
            ok: true,
            port: parsed.options.port,
            browser: snapshot.browser,
            tabs: harness.listTabs()
          },
          parsed.options.pretty
        );
        return 0;
      }

      case 'open-tab': {
        const tab = await harness.openTab(parsed.url!, parsed.focus);
        writeJson(
          {
            ok: true,
            tab
          },
          parsed.options.pretty
        );
        return 0;
      }

      case 'discover-tools': {
        const discovered = await harness.discoverTools(parsed.tabId!);
        writeJson(
          {
            ok: true,
            tabId: parsed.tabId,
            tools: discovered.tools
          },
          parsed.options.pretty
        );
        return 0;
      }

      case 'call-tool': {
        const result = await harness.callTool(
          parsed.tabId!,
          parsed.toolName!,
          parsed.toolArgs
        );

        writeJson(
          {
            ok: true,
            tabId: parsed.tabId,
            toolName: parsed.toolName,
            result
          },
          parsed.options.pretty
        );
        return 0;
      }
    }
  } catch (error) {
    writeJson(
      {
        ok: false,
        error: error instanceof Error ? error.message : String(error)
      },
      parsed.options.pretty
    );
    return 1;
  } finally {
    await harness.close();
  }
}

void run().then((code) => {
  process.exit(code);
});
